File: 904-leaf-similar-trees/leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        // Explicitly define the type so it can reference itself
        std::function<void(TreeNode*, vector<int>&)> getLeaves = [&](TreeNode* root, vector<int>& leaves) {
            if (!root) return;
            if (!root->left && !root->right) {
                leaves.push_back(root->val); // Note: use root->val, not root.val
                return;
            }
            getLeaves(root->left, leaves);
            getLeaves(root->right, leaves);
        };

        vector<int> leaves1, leaves2;
        getLeaves(root1, leaves1);
        getLeaves(root2, leaves2);
        
        return leaves1 == leaves2;
    };
};

File: 904-leaf-similar-trees/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        
        def get_leaf_values(node):
            if not node: return []
            if not node.left and not node.right: # It's a leaf node
                return [node.val]
            # recursively get leaves from left and right subtrees
            return get_leaf_values(node.left) + get_leaf_values(node.right)

        # get leaf sequences for both trees
        return get_leaf_values(root1) == get_leaf_values(root2)


        